### PR TITLE
Add `pkg/log` package and cleanup logging around the codebase

### PR DESCRIPTION
**Pull request by neoaggelos**
### Summary

Add `pkg/log` with a standard logging interface and document how we should create logs across the codebase.

Start from https://github.com/canonical/k8s-snap/compare/KU-1080/logging?expand=1#diff-3e6c5854477b8113c122f67474e7fba607496ef206b4a2a0761cdde6cc26b9b9

Replace usage of the standard library "log" with `pkg/log` 

### Notes

This is pre-requisite work for the CSRSigning controller for certificate refreshes.

### CODE REVIEW

File: src/k8s/pkg/k8sd/api/cluster_remove.go [hunk 1]
```diff
@@ -32,29 +32,32 @@ func (e *Endpoints) postClusterRemove(s *state.State, r *http.Request) response.
 		defer cancel()
 	}
 
+	log := log.FromContext(s.Context).WithValues("name", req.Name)
```
Comment (bschimke95) [inline src/k8s/pkg/k8sd/api/cluster_remove.go:35]: nit: `WithName`?

File: src/k8s/pkg/k8sd/api/impl/k8sd.go [hunk 2]
```diff
@@ -46,22 +45,15 @@ func GetLocalNodeStatus(ctx context.Context, s *state.State, snap snap.Snap) (ap
...
-			log.Printf("Failed to check if node is control-plane. This is expected if the node is in a joining/leaving phase. %v", err)
-			clusterRole = apiv1.ClusterRoleUnknown
-		} else {
-			if node != nil {
-				return *node, nil
-			}
-		}
-
+	} else if node, err := nodeutil.GetControlPlaneNode(ctx, s, s.Name()); err != nil {
+		clusterRole = apiv1.ClusterRoleUnknown
```
Comment (bschimke95) [inline src/k8s/pkg/k8sd/api/impl/k8sd.go:52]: why did you remove the additional logging?

File: src/k8s/pkg/client/helm/client.go [hunk 3]
```diff
@@ -31,18 +31,21 @@ func NewClient(manifestsBaseDir string, restClientGetter func(string) genericcli
 	}
 }
 
-func (h *client) newActionConfiguration(namespace string) (*action.Configuration, error) {
+func (h *client) newActionConfiguration(ctx context.Context, namespace string) (*action.Configuration, error) {
 	actionConfig := new(action.Configuration)
 
-	if err := actionConfig.Init(h.restClientGetter(namespace), namespace, "", log.Printf); err != nil {
+	log := log.FromContext(ctx).WithName("helm")
```
Comment (louiseschmidtgen) [inline src/k8s/pkg/client/helm/client.go:37]: Neat!

File: src/k8s/pkg/k8sd/controllers/update_node_configuration.go [hunk 4]
```diff
@@ -58,6 +58,8 @@ func (c *UpdateNodeConfigurationController) retryNewK8sClient(ctx context.Contex
 func (c *UpdateNodeConfigurationController) Run(ctx context.Context, getClusterConfig func(context.Context) (types.ClusterConfig, error)) {
 	c.waitReady()
 
+	log := log.FromContext(ctx).WithValues("controller", "updatenodeconfiguration")
```
Comment (bschimke95) [inline src/k8s/pkg/k8sd/controllers/update_node_configuration.go:61]: nit: snake case `update_node_configuration` would probably be easier to read.

File: src/k8s/pkg/k8sd/api/impl/k8sd.go (hunk 2 above)
Comment (louiseschmidtgen) [inline src/k8s/pkg/k8sd/api/impl/k8sd.go:52]: Is the log here no longer necessary?

Comment (neoaggelos) [reply src/k8s/pkg/k8sd/api/impl/k8sd.go:52]: yes, removed log lines that did not provide any useful context here and in some other places

File: src/k8s/pkg/proxy/apiserver.go [hunk 5]
```diff
@@ -68,18 +72,19 @@ func (p *APIServerProxy) watchForNewEndpoints(ctx context.Context, cancel func()
 		newEndpoints, err := getKubernetesEndpoints(ctx, p.KubeconfigFile)
 		switch {
 		case err != nil:
-			log.Println(fmt.Errorf("failed to retrieve kubernetes endpoints: %w", err))
+			log.Error(err, "Failed to retrieve Kubernetes endpoints")
 			continue
 		case len(newEndpoints) == 0:
-			log.Println("warning: empty list of endpoints, skipping update")
+			log.Info("Warning: empty list of endpoints, skipping update")
```
Comment (bschimke95) [inline src/k8s/pkg/proxy/apiserver.go:78]: should this be a warning level message instead of prefix this with "Warning"?

**Review by bschimke95 — APPROVED** [review]
LGTM, minor comments.
Great work!

File: src/k8s/pkg/k8sd/api/cluster_remove.go (hunk 1 above)
Comment (neoaggelos) [reply src/k8s/pkg/k8sd/api/cluster_remove.go:35]: note: we are adding the node name as `name=foo`. `log.WithName("test")` would give us a field with `logger="k8sd.test"` which is not what we want here

File: src/k8s/pkg/proxy/proxy.go [hunk 6]
```diff
@@ -41,10 +42,15 @@ func startProxy(ctx context.Context, listenURL string, endpointURLs []string) er
 		MonitorInterval: time.Minute,
 	}
 
-	log.Println("Starting proxy at", listenURL)
+	log := log.FromContext(ctx).WithValues(
```
Comment (louiseschmidtgen) [inline src/k8s/pkg/proxy/proxy.go:45]: This is nice!

**Review by louiseschmidtgen — APPROVED** [review]
Great work, loving the improvement on our logs :)

File: src/k8s/pkg/proxy/apiserver.go (hunk 5 above)
Comment (neoaggelos) [reply src/k8s/pkg/proxy/apiserver.go:78]: Logs are either `.Info()` (with log levels) or `.Error()` (always printed)

File: src/k8s/pkg/k8sd/api/impl/k8sd.go (hunk 2 above)
Comment (bschimke95) [reply src/k8s/pkg/k8sd/api/impl/k8sd.go:52]: I would keep around the comment on when this might happen. Let's remove the logs though.
